src/common/sys_lmdb.c

```diff
@@ -0,0 +1,517 @@
+	D_FREE(db_key.mv_data);
+
+out:
+	rc = mdb_error2daos_error(rc);
+	if (end_tx)
+		lmm_db_tx_end(db, rc);
+	return rc;
+}
+
+static int
+lmm_db_delete(struct sys_db *db, char *table, d_iov_t *key)
+{
+	MDB_val			 db_key;
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	int			 rc;
+	bool			 end_tx = false;
+
+	if (ldb->db_txn == NULL) {
+		rc = lmm_db_tx_begin(db);
+		if (rc)
+			return rc;
+		end_tx = true;
+	}
+
+	rc = lmm_db_generate_key(table, key, &db_key);
+	if (rc)
+		goto out;
+
+	rc = mdb_del(ldb->db_txn, ldb->db_dbi, &db_key, NULL);
+	D_FREE(db_key.mv_data);
+
+out:
+	rc = mdb_error2daos_error(rc);
+	if (end_tx)
+		lmm_db_tx_end(db, rc);
+	return rc;
+}
+
+static int
+lmm_db_traverse(struct sys_db *db, char *table, sys_db_trav_cb_t cb, void *args)
+{
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	MDB_cursor		*cursor;
+	MDB_val			 db_key, db_data;
+	int			 rc;
+	d_iov_t			 key;
+	int			 table_len = strnlen(table, MAX_SMD_TABLE_LEN);
+
+	D_ASSERT(ldb->db_txn == NULL);
+	rc = mdb_txn_begin(ldb->db_env, NULL, MDB_RDONLY, &ldb->db_txn);
+	if (rc)
+		return mdb_error2daos_error(rc);
+
+	rc = mdb_cursor_open(ldb->db_txn, ldb->db_dbi, &cursor);
+	if (rc)
+		goto tx_end;
+
+	while ((rc = mdb_cursor_get(cursor, &db_key, &db_data, MDB_NEXT)) == 0) {
+		if (strncmp(db_key.mv_data, table, table_len) != 0)
+			continue;
+
+		rc = lmm_db_unpack_key(table, &key, &db_key);
+		if (rc)
+			break;
+
+		d_iov_set(&key, db_key.mv_data, db_key.mv_size);
+		rc = cb(db, table, &key, args);
+		D_FREE(key.iov_buf);
+		if (rc)
+			break;
+	}
+	/* reach end */
+	if (rc == MDB_NOTFOUND)
+		rc = 0;
+
+	mdb_cursor_close(cursor);
+tx_end:
+	mdb_txn_abort(ldb->db_txn);
+	ldb->db_txn = NULL;
+	return mdb_error2daos_error(rc);
+}
+
+static int
+lmm_db_tx_begin(struct sys_db *db)
+{
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	int			 rc;
+
+	D_ASSERT(ldb->db_txn == NULL);
+	rc = mdb_txn_begin(ldb->db_env, NULL, 0, &ldb->db_txn);
+
+	return mdb_error2daos_error(rc);
+}
+
+static int
+lmm_db_tx_end(struct sys_db *db, int rc)
+{
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	MDB_txn			*txn = ldb->db_txn;
+
+	D_ASSERT(txn != NULL);
+	ldb->db_txn = NULL;
+
+	if (rc) {
+		mdb_txn_abort(txn);
+		return rc;
+	}
+
+	rc = mdb_txn_commit(txn);
+
+	return mdb_error2daos_error(rc);
+}
+
+static void
+lmm_db_lock(struct sys_db *db)
+{
+	ABT_mutex_lock(db2lmm(db)->db_lock);
+}
+
+static void
+lmm_db_unlock(struct sys_db *db)
+{
+	ABT_mutex_unlock(db2lmm(db)->db_lock);
+}
+
+/** Finalize system DB of VOS */
+void
+lmm_db_fini(void)
+{
+	if (lmm_db.db_lock)
+		ABT_mutex_free(&lmm_db.db_lock);
+	if (lmm_db.db_file) {
+		if (lmm_db.db_env)
+			mdb_env_close(lmm_db.db_env);
+		if (lmm_db.db_destroy_db)
+			lmm_db_unlink(&lmm_db.db_pub);
+		mdb_dbi_close(lmm_db.db_env, lmm_db.db_dbi);
+		free(lmm_db.db_file);
+	}
+
+	free(lmm_db.db_path);
+	memset(&lmm_db, 0, sizeof(lmm_db));
+}
+
+int
+lmm_db_init_ex(const char *db_path, const char *db_name, bool force_create, bool destroy_db_on_fini)
```

Comment:
Maybe we could improve the src/bio/smd/tests/smd_ut.c to make the backend sysdb configurable (specify backend sysdb by argument)? So that we could run SMD unit tests over lmdb based sysdb.